siliconcompiler/schema/utils.py

```diff
@@ -0,0 +1,83 @@
+# SC dependencies outside of its directory, since it may be used by tool drivers
+# that have isolated Python environments.
+
+import os
+import re
+import sys
+
+PACKAGE_ROOT = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..')
+
+def escape_val_tcl(val, typestr):
+    '''Recursive helper function for converting Python values to safe TCL
+    values, based on the SC type string.'''
+    if val is None:
+        return ''
+    elif typestr.startswith('('):
+        # Recurse into each item of tuple
+        subtypes = typestr.strip('()').split(',')
+        valstr = ' '.join(escape_val_tcl(v, subtype.strip())
+                          for v, subtype in zip(val, subtypes))
+        return f'[list {valstr}]'
+    elif typestr.startswith('['):
+        # Recurse into each item of list
+        subtype = typestr.strip('[]')
+        valstr = ' '.join(escape_val_tcl(v, subtype) for v in val)
+        return f'[list {valstr}]'
+    elif typestr == 'bool':
+        return 'true' if val else 'false'
+    elif typestr == 'str':
+        # Escape string by surrounding it with "" and escaping the few
+        # special characters that still get considered inside "". We don't
+        # use {}, since this requires adding permanent backslashes to any
+        # curly braces inside the string.
+        # Source: https://www.tcl.tk/man/tcl8.4/TclCmd/Tcl.html (section [4] on)
+        escaped_val = (val.replace('\\', '\\\\') # escape '\' to avoid backslash substition (do this first, since other replaces insert '\')
+                            .replace('[', '\\[')   # escape '[' to avoid command substition
+                            .replace('$', '\\$')   # escape '$' to avoid variable substition
+                            .replace('"', '\\"'))  # escape '"' to avoid string terminating early
+        return '"' + escaped_val + '"'
+    elif typestr in ('file', 'dir'):
+        # Replace $VAR with $env(VAR) for tcl
+        val = re.sub(r'\$(\w+)', r'$env(\1)', val)
+        # Same escapes as applied to string, minus $ (since we want to resolve env vars).
+        escaped_val = (val.replace('\\', '\\\\') # escape '\' to avoid backslash substition (do this first, since other replaces insert '\')
+                            .replace('[', '\\[')   # escape '[' to avoid command substition
+                            .replace('"', '\\"'))  # escape '"' to avoid string terminating early
+        return '"' +  escaped_val + '"'
+    else:
+        # floats/ints just become strings
+        return str(val)
+
+def trim(docstring):
+    '''Helper function for cleaning up indentation of docstring.
+
+    This is important for properly parsing complex RST in our docs.
+
+    Source:
+    https://www.python.org/dev/peps/pep-0257/#handling-docstring-indentation'''
+    if not docstring:
+        return ''
+    # Convert tabs to spaces (following the normal Python rules)
+    # and split into a list of lines:
+    lines = docstring.expandtabs().splitlines()
```

Comment:
If you call `.strip()` before `.splitlines()`, you might be able to skip the logic below which removes leading/trailing newlines

Reply:
Seems like a good catch - I'll probably keep the logic as-is since it came verbatim from the linked PEP, though!